packages/native/src/useLinking.tsx

```diff
@@ -123,15 +126,26 @@ const createMemoryHistory = () => {
     go(n: number) {
       interrupt();
 
+      // To guard against unexpected navigation out of the app we will assume that browser history is only as deep as the length of our memory
+      // history. If we don't have an item to navigate to then update our index and navigate as far as we can without taking the user out of the app.
+      const nextIndex = index + n;
+      const lastItemIndex = items.length - 1;
+      if (n < 0 && !items[nextIndex]) {
+        // Attempted to navigate beyond the first index. Negating the current index will align the browser history with the first item.
+        n = -index;
+        index = 0;
+      } else if (n > 0 && nextIndex > lastItemIndex) {
+        // Attempted to navigate past the last index. Calculate how many indices away from the last index and go there.
+        n = lastItemIndex - index;
+        index = lastItemIndex;
+      } else {
+        index = nextIndex;
+      }
+
       if (n === 0) {
         return;
       }
 
-      // We shouldn't go back more than the 0 index (otherwise we'll exit the page)
-      // Or forward more than the available index (or the app will crash)
-      index =
-        n < 0 ? Math.max(index - n, 0) : Math.min(index + n, items.length - 1);
```

Comment:
@satya164 I _think_ this logic is incorrect. If we set have a negative `n` delta (i.e. navigating back) then we will set the index to either `0` or `index - delta`, but since `delta` is negative we should be adding `index + n` not `index - n` ?

e.g. say we have 2 items and our index is `1` and our delta is `n = -1` then our new index would be `2` or `0` when the correct index should be `0`.

![Screen Shot 2022-05-22 at 6 57 34 AM](https://user-images.githubusercontent.com/32969087/169706735-ee012352-1424-4d9c-9a1d-a8966d1a0e49.png)

Comment also seems out of date as setting the index does not prevent us from navigating outside the bounds since we are always calling `window.history.go(n)` with an unmodified delta.

https://github.com/react-navigation/react-navigation/blob/1e1bc84b64ca0f740e2d1b7ff531d64efdb56887/packages/native/src/useLinking.tsx#L210

Reply:
@marcaaron yes, you're right this logic is incorrect as `delta` is negative and we should be adding `index + n` i figured it out as well during [debugging](https://github.com/Expensify/App/issues/7618#:~:text=there%20are%20a%20few%20other%20issues%20that%20i%20have%20found%20in%20react%2Dnavigation). Was about to raise a PR to fix this after the other [fix](https://github.com/react-navigation/react-navigation/pull/10600). 

Thank you for raising this PR 🙇